contracts/validators/AuthV2Validator.sol

```diff
@@ -0,0 +1,173 @@
+     * @param c πc element of the groth16 proof.
+     * @param data Request query data of the credential to verify.
+     * @param sender Sender of the proof.
+     * @return Array of key to public input index as result.
+     */
+    function verify(
+        uint256[] memory inputs,
+        uint256[2] memory a,
+        uint256[2][2] memory b,
+        uint256[2] memory c,
+        bytes calldata data,
+        address sender
+    ) public view override returns (ICircuitValidator.KeyToInputIndex[] memory) {
+        ICircuitValidator.KeyToInputIndex[]
+            memory keyToInputIndex = new ICircuitValidator.KeyToInputIndex[](1);
+        if (inputs.length == 0) {
+            uint256 userId = GenesisUtils.calcIdFromEthAddress(
+                _getState().getDefaultIdType(),
+                sender
+            );
+
+            keyToInputIndex[0] = ICircuitValidator.KeyToInputIndex({
+                key: "userID",
+                inputIndex: userId
+            });
+            return keyToInputIndex;
+        }
+
+        _verifyMain(inputs, a, b, c, data, sender, IState(getStateAddress()));
+        keyToInputIndex[0] = ICircuitValidator.KeyToInputIndex({
+            key: "userID",
+            inputIndex: inputs[0]
+        });
+        return keyToInputIndex;
+    }
+
+    /**
+     * @dev Verify the groth16 proof and check the request query data
+     * @param zkProof Proof packed as bytes to verify.
+     * @param data Request query data of the credential to verify.
+     * @param sender Sender of the proof.
+     * @param stateContract State contract to get identities and gist states to check.
+     * @return Array of public signals as result.
+     */
+    function verifyV2(
+        bytes calldata zkProof,
+        bytes calldata data,
+        address sender,
+        IState stateContract
+    ) public view override returns (ICircuitValidator.Signal[] memory) {
+        if (zkProof.length == 0) {
+            uint256 userId = GenesisUtils.calcIdFromEthAddress(
+                _getState().getDefaultIdType(),
+                sender
+            );
+            ICircuitValidator.Signal[] memory signals = new ICircuitValidator.Signal[](1);
+            signals[0] = ICircuitValidator.Signal({name: "userID", value: userId});
+            return signals;
+        }
+
+        (
+            uint256[] memory inputs,
+            uint256[2] memory a,
+            uint256[2][2] memory b,
+            uint256[2] memory c
+        ) = abi.decode(zkProof, (uint256[], uint256[2], uint256[2][2], uint256[2]));
+
+        return _verifyMain(inputs, a, b, c, data, sender, stateContract);
+    }
+
+    function _verifyZKP(
+        uint256[] memory inputs,
+        uint256[2] memory a,
+        uint256[2][2] memory b,
+        uint256[2] memory c
+    ) internal view {
+        IVerifier verifier = getVerifierByCircuitId(CIRCUIT_ID);
+        require(verifier != IVerifier(address(0)), "Verifier address should not be zero");
```

Comment:
This check is something which can be done in initialiser and one time only. Yes, you'll need to change `_initDefaultStateVariables` method in the base class but it's not a big work we'll finally stop repeating it in every validator.

Separate PR is better for it, of course.